File: packages/ts/src/proto.ts
// Shared loader: parse ../proto/tools.proto at runtime. The dynamic loader is
// untyped by nature, so we declare the contract shapes ourselves and cast the
// loaded descriptor to them — giving the rest of the code real type-safety.
import * as grpc from '@grpc/grpc-js';
import * as protoLoader from '@grpc/proto-loader';
import { fileURLToPath } from 'node:url';
import path from 'node:path';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

export const PROTO_PATH = path.resolve(__dirname, '../../proto/tools.proto');
export const SAMPLE_TOOLS_PATH = path.resolve(__dirname, '../../proto/tools.sample.json');

export interface Tool {
  name: string;
  description: string;
  input_schema: string;
}

export interface ToolListResponse {
  tools: Tool[];
}

const packageDefinition = protoLoader.loadSync(PROTO_PATH, {
  keepCase: true,
  longs: String,
  enums: String,
  defaults: true,
  oneofs: true,
});

const protoDescriptor = grpc.loadPackageDefinition(packageDefinition) as unknown as {
  api: { DiscoveryService: grpc.ServiceClientConstructor };
};

export const DiscoveryService = protoDescriptor.api.DiscoveryService;

File: packages/py/proto.py
"""Shared loader: parse ../proto/tools.proto at runtime.

`grpc.protos()` / `grpc.services()` compile the .proto in-memory (no committed
*_pb2.py stubs), mirroring @grpc/proto-loader on the JS side. They require the
`grpcio-tools` package and the proto's directory on sys.path.
"""
import os
import sys

import grpc

PROTO_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "proto"))
SAMPLE_TOOLS_PATH = os.path.join(PROTO_DIR, "tools.sample.json")

# Make "tools.proto" importable by the dynamic loader.
if PROTO_DIR not in sys.path:
    sys.path.insert(0, PROTO_DIR)

# tools_pb2 -> messages (EmptyRequest, Tool, ToolListResponse)
# tools_pb2_grpc -> DiscoveryServiceStub / Servicer / registration helper
tools_pb2 = grpc.protos("tools.proto")
tools_pb2_grpc = grpc.services("tools.proto")
